compiler/src/dotty/tools/dotc/core/Names.scala

```diff
@@ -167,35 +153,107 @@ object Names {
+        derivedNames.get(info)
+    }
+
+    private def putDerived(info: NameInfo, name: DerivedTermName): name.type = {
+      derivedNames match {
+        case derivedNames: SimpleMap[NameInfo, DerivedTermName] @unchecked =>
+          if (derivedNames.size < 4)
+            this.derivedNames = derivedNames.updated(info, name)
+          else {
+            val newMap = new HashMap[NameInfo, DerivedTermName]
+            derivedNames.foreachBinding(newMap.put(_, _))
+            newMap.put(info, name)
+            this.derivedNames = newMap
+          }
+        case derivedNames: HashMap[NameInfo, DerivedTermName] @unchecked =>
+          derivedNames.put(info, name)
+      }
+      name
+    }
+
+    /** Return derived name with given `info` and the current
+     *  name as underlying name.
+     */
+    def derived(info: NameInfo): TermName = {
+      def addIt() = synchronized {
+        getDerived(info) match {
+          case null => putDerived(info, new DerivedTermName(this, info))
+          case derivedName => derivedName
+        }
+      }
+
+      val ownKind = this.info.kind
+      if (ownKind > info.kind)
+        underlying.derived(info).derived(this.info)
+      else if (ownKind == info.kind)
+        if (info.oneOfAKind) {
+          assert(info == this.info)
+          this
+        }
+        else addIt()
+      else addIt()
+    }
+
+    def is(kind: NameInfo.Kind): Boolean = {
+      val ownKind = info.kind
+      ownKind == kind || ownKind > kind && underlying.is(kind)
+    }
 
     override def hashCode: Int = start
 
     override protected[this] def newBuilder: Builder[Char, Name] = termNameBuilder
+  }
 
-    def fromChars(cs: Array[Char], offset: Int, len: Int): TermName = termName(cs, offset, len)
+  class SimpleTermName(val start: Int, val length: Int, @sharable private[Names] var next: SimpleTermName) extends TermName {
+    // `next` is @sharable because it is only modified in the synchronized block of termName.
   }
 
-  class TypeName(val start: Int, val length: Int, val toTermName: TermName) extends Name {
+  class TypeName(val toTermName: TermName) extends Name {
+    def start = toTermName.start
+    override def length = toTermName.length
+
     type ThisName = TypeName
     def isTypeName = true
     def isTermName = false
     def toTypeName = this
     def asTypeName = this
     def asTermName = throw new ClassCastException(this + " is not a term name")
 
+    def fromName(name: Name): TypeName = name.toTypeName
+
     override def hashCode: Int = -start
 
     override protected[this] def newBuilder: Builder[Char, Name] =
       termNameBuilder.mapResult(_.toTypeName)
+  }
 
-    def fromChars(cs: Array[Char], offset: Int, len: Int): TypeName = typeName(cs, offset, len)
+  /** A term name that's derived from an `underlying` name and that
+   *  adds `info` to it.
+   */
+  class DerivedTermName(override val underlying: TermName, override val info: NameInfo)
```

Comment:
Instead of having a `NameInfo` object as parameter, what about a `NameFlagSet` ?